astropy/coordinates/sky_coordinate.py

```diff
@@ -1205,46 +1213,80 @@ def _parse_coordinate_arg(coords, frame, units):
         values = coords.transpose()  # Iterates over repr attrs
 
     elif isinstance(coords, (collections.Sequence, np.ndarray)):
-        # Handles generic list-like input.
+        # Handles list-like input.
 
-        # First turn into a list of lists like [[v1_0, v2_0, v3_0], ... [v1_N, v2_N, v3_N]]
+        # First turn into a list of lists like [[v1_0, v2_0, v3_0], ... [v1_N, v2_N, v3_N]],
+        # and check if there are any coordinate objects mixed in
         vals = []
-        for ii, coord in enumerate(coords):
+        hascoordobjs = False
```

Comment:
What about reducing the complexity a bit with
```
coord_classes = (SkyCoord, BaseCoordinateFrame, BaseRepresentation)
if any(isinstance(coord, coord_classes) for coord in coords):
    # This parsing path is used ...
else:
    vals = []
    for coord in cords:
        if isinstance(coord, six.string_types):
            ...
```
You do pay a penalty of sometimes having the top-level loop run twice, but I think in practice the overhead for just the looping is quite small compared to everything else.   The benefit is that the code logic is more distinct.  And if it is possible to refactor the two code paths as I suggested above then things get even more clear.

I definitely agree this part of the code can use some TLC.  BTW, worth noting #2920 which is fiddling in the same function.  Will likely conflict but not in a difficult way.